File: app/src/main/java/com/mancode/financetracker/database/entity/TransactionEntity.java
package com.mancode.financetracker.database.entity;

import com.mancode.financetracker.database.converter.DateConverter;

import java.util.Date;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.PrimaryKey;
import androidx.room.TypeConverters;

/**
 * Created by Manveru on 25.01.2018.
 */

@Entity(tableName = "transactions",
        foreignKeys = {
                @ForeignKey(
                        entity = AccountEntity.class,
                        parentColumns = "_id",
                        childColumns = "transaction_account"),
                @ForeignKey(
                        entity = CategoryEntity.class,
                        parentColumns = "_id",
                        childColumns = "transaction_category")
        }
)
@TypeConverters(DateConverter.class)
public class TransactionEntity {

    public static final int TYPE_INCOME = 1;
    public static final int TYPE_OUTCOME = -1;

    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "_id")
    public int id;

    @ColumnInfo(name = "transaction_date")
    public Date date;

    @ColumnInfo(name = "transaction_type")
    public int type;

    @ColumnInfo(name = "transaction_description")
    public String description;

    @ColumnInfo(name = "transaction_value")
    public double value;

    @ColumnInfo(name = "transaction_account")
    public int accountId;

    @ColumnInfo(name = "transaction_category")
    public int categoryId;

    public TransactionEntity(int id, Date date, int type, String description,
                             double value, int accountId, int categoryId) {
        this.id = id;
        this.date = date;
        this.type = type;
        this.description = description;
        this.value = value;
        this.accountId = accountId;
        this.categoryId = categoryId;
    }
}

File: app/src/main/java/com/mancode/financetracker/database/entity/TransactionEntity.kt
package com.mancode.financetracker.database.entity

import androidx.room.*
import com.mancode.financetracker.database.converter.DateConverter
import org.threeten.bp.LocalDate

/**
 * Created by Manveru on 25.01.2018.
 */

@Entity(tableName = "transactions",
        foreignKeys = [
            ForeignKey(entity = AccountEntity::class,
                    parentColumns = ["_id"],
                    childColumns = ["transaction_account"]),
            ForeignKey(entity = CategoryEntity::class,
                    parentColumns = ["_id"],
                    childColumns = ["transaction_category"])])
@TypeConverters(DateConverter::class)
data class TransactionEntity(
        @PrimaryKey(autoGenerate = true)
        @ColumnInfo(name = "_id")
        val id: Int,
        @ColumnInfo(name = "transaction_date")
        var date: LocalDate,
        @ColumnInfo(name = "transaction_type")
        var type: Int,
        @ColumnInfo(name = "transaction_description")
        var description: String,
        @ColumnInfo(name = "transaction_value")
        var value: Double,
        @ColumnInfo(name = "transaction_flags", defaultValue = "0")
        var flags: Int,
        @ColumnInfo(name = "transaction_account")
        var accountId: Int,
        @ColumnInfo(name = "transaction_category")
        var categoryId: Int) {

    companion object {
        const val TYPE_INCOME = 1
        const val TYPE_OUTCOME = -1

        // FLAGS
        const val BOOKMARKED = 1
    }
}
